src/vmprof_common.h

```diff
@@ -108,9 +114,52 @@ static int opened_profile(char *interp_name, int memory)
  
 PyThreadState* get_current_thread_state(void)
 {
+    PyThreadState *tstate;
 #if PY_MAJOR_VERSION >= 3
-    return (PyThreadState*)_Py_atomic_load_relaxed(&_PyThreadState_Current);
+    tstate = (PyThreadState*)_Py_atomic_load_relaxed(&_PyThreadState_Current);
 #else
-    return _PyThreadState_Current;
+    tstate = _PyThreadState_Current;
 #endif
+
+    if (tstate == NULL) {
```

Comment:
I understand, I've tried using good APIs from CPython, unfortunately this is not safe (and will end up in crashes), since you can't walk this list without holding a lock - it's not a given that some other thread is not running at the same time (even if your thread is suspended right now)